theano/gpuarray/tests/test_dnn.py

```diff
@@ -2298,3 +2298,221 @@ class Cudnn_grouped_conv(Grouped_conv_noOptim):
+        # It is equivalent to the following numpy code:
+        #  x_t, y_t = np.meshgrid(np.linspace(-1, 1, width),
+        #                         np.linspace(-1, 1, height))
+        #  ones = np.ones(np.prod(x_t.shape))
+        #  grid = np.vstack([x_t.flatten(), y_t.flatten(), ones])
+        # It is implemented in Theano instead to support symbolic grid sizes.
+        # Note: If the image size is known at layer construction time, we could
+        # compute the meshgrid offline in numpy instead of doing it dynamically
+        # in Theano. However, it hardly affected performance when we tried.
+        x_t = T.dot(T.ones((height, 1)),
+                    _linspace(-1.0, 1.0, width).dimshuffle('x', 0))
+        y_t = T.dot(_linspace(-1.0, 1.0, height).dimshuffle(0, 'x'),
+                    T.ones((1, width)))
+
+        x_t_flat = x_t.reshape((1, -1))
+        y_t_flat = y_t.reshape((1, -1))
+        ones = T.ones_like(x_t_flat)
+        grid = T.concatenate([x_t_flat, y_t_flat, ones], axis=0)
+        return grid
+
+    # Generate random set of RGB images with 256x256 resolution (pixel values in [0, 255])
+    img_dims = (10, 256, 256, 3)  # images are usually NHWC
+    img = np.random.randint(low=0, high=256, size=img_dims)
+    # Convert from NHWC to NCHW
+    img = np.transpose(img, axes=(0, 3, 1, 2)).astype(theano.config.floatX)
+    scale_height = 0.25
+    scale_width = 0.75
+
+    # Transformation matrix
+    theta = [[-1, 0, 0],
+             [0, -1, 0]]
+
+    transform = np.asarray(img_dims[0] * [theta], dtype=theano.config.floatX)
+
+    # Create symbolic variables for inputs and transformations
+    t_img = T.tensor4('img')
+    t_theta = T.tensor3('theta')
+
+    st_dnn = dnn.dnn_spatialtf(t_img, t_theta, scale_height=scale_height,
+                               scale_width=scale_width)
+    st_dnn_func = theano.function([t_img, t_theta], st_dnn)
+    # Check if function graph contains the spatial transformer's grid and sampler Ops
+    assert any([isinstance(node.op, dnn.GpuDnnTransformerGrid)
+                for node in st_dnn_func.maker.fgraph.toposort()])
+    assert any([isinstance(node.op, dnn.GpuDnnTransformerSampler)
+                for node in st_dnn_func.maker.fgraph.toposort()])
+
+    img_out_gpu = st_dnn_func(img, transform)
+    img_out = np.asarray(img_out_gpu)
+
+    # Setup CPU Op
+    st_cpu = spatialtf_cpu(t_theta, t_img, scale_height, scale_width, 'nearest')
+    st_cpu_func = theano.function([t_theta, t_img], [st_cpu], mode=mode_without_gpu)
+    res, = st_cpu_func(transform, img)
+
+    utt.assert_allclose(img_out, res, rtol=1e-2, atol=1e-2)
+
+
+def test_dnn_spatialtf_grad():
+    if not dnn.dnn_available(test_ctx_name):
+        raise SkipTest(dnn.dnn_available.msg)
+
+    utt.seed_rng()
+
+    # Generate random set of RGB images with 256x256 resolution (pixel values in [0, 255])
+    img_dims = (1, 256, 256, 3)  # images are usually NHWC
```

Comment:
It seems gradient tests take time because inputs are quuite big. Use smaller images (e.g. width 10 and height 10).